rfcs/20181029-more-flexible-concurrency.md

```diff
@@ -0,0 +1,96 @@
+as well as multiple independent tasks all transforming parts of the same data
+chunk. A major constraint is our processes must restrict the number of
+processing threads they use to a number assigned to them at process start. This
+is needed since we use computing resources shared by multiple users with each
+user assigned a certain number of computing cores they are allowed to use.
+(NOTE: a small number of primarily 'waiting' threads is permissible, as long as
+the processing time used by those threads is negligible.)
+
+We use the Intel Thread Building Blocks (TBB) library to manage a thread pool
+of processing tasks. We tell TBB the total number of cores it is allowed to use
+and that sets our processing limits for the process. In order to use TensorFlow
+based inferences in this system, we modified the Session class to not use any
+threads. This allowed the inference to be evaluated completely on the TBB
+managed thread which is running the task needed the inference is using. This
+works, but we would like to be able to do better. There are cases where the
+total number of TBB tasks available to the process at a given time is less than
+the number of threads available to TBB. Under those cases, we would like to be
+able to use the underutilized threads to speed up the TensorFlow inference
+calculation. This could be accomplished if TensorFlow had a mechanism to allow
+alternative concurrency engines to be 'plugged into' the system. We would then
+use those APIs to implement a TBB based concurrency engine. Such a facility
+could also be used to allow OpenMP, another common scientific concurrency
+engine, to be developed as well.
+
+To be clear - we are not asking for TBB support, but for a way to add it
+ourselves as a pull request. This would also enable other users to write code
+to support, for example, OpenMP.
+
+## Design Proposal
+
+The most important features we require are:
+
+1. Control the ability to run inter-op closures executed using the
+`default_runner`.
+2. We don't need to make changes to `EigenCPUDevice`, and are not concerned
+about intra-op parallelism.
+3. We want some control before the session-run completes, to handle `taskArena`
+or `taskGroup` cleanup.
+
+We believe it is possible to leverage the newly introduced `RunHandlerPool`
+instead of a new session Class. At a high level this would introduce a new
+implementation of `RunHandlerPool` that works with TBB instead of Eigen, and
+returns a `TBBRunHandler` on `RunHandlerPool::Get()`. We may want to make the
+existing `RunHandler` an interface with a pure virtual method for
+`ScheduleInterOpClosure()`. This would produce our required control over
+running inter-op closures. This interface currently doesn't support intra-op
+parallelism, so we assume our second feature is satisfied in this case.
+
+Each `RunHandler` object is active only during a session run, so we should get
+control before the destruction.
+
+## Detailed Design
```

Comment:
Can you propose here what kind of API entry point would you be interested in? For example, a way to configure session.run using the C API to use your alternative thread pool, etc.

Reply:
Maybe the clearest way to answer that question is to show the modifications we made for TF 1.X to get this behavior:

https://gist.github.com/gnperdue/d7af021a2e9e0467aec7fbbc8a0c3137

Reply:
Very interesting. If you could add more details to this document about the concept of a task arena and how it relates to the other things in tf this would be good.

Reply:
A member of the CMS team took a look at this and had the following comments:

```
I finally had a chance to look at the comments and back at the CMS code. The line 
'
3. We want some control before the session-run completes, to handle `taskArena`
or `taskGroup` cleanup.
'

Is probably too specific. What we really need is a way to 
- Setup up state for each call to Session::Run
	(this is where we created the task arena and task group)
- Use that state as part of the args given to 'executor->RunAsync'
	(this was part of the change to  Session::SchedClosure to get access to the task arena and task group)
- Use that state as part of the Session::WaitForNotification to allow the thread to be used for processing as well as to wait for all outstanding work to finish. 
	(this was part of the different Session::WaitForNotification which used the task_arena and task_group).
```

Does that help?